utils/dialog/dialog.man
.TH DIALOG 1 "10 January 1994"
.SH NAME
dialog \- display dialog boxes from shell scripts
.SH SYNOPSIS
.B dialog --clear
.br
.BI "dialog --create-rc " file
.br
.B dialog
[
.B \-\-title
.I title
]
[
.B \-\-clear
]
.B box-options
.SH DESCRIPTION
.B Dialog
is a program that will let you to present a variety of questions or
display messages using dialog boxes from a shell script. Currently,
these types of dialog boxes are implemented: 
.LP
.BR yes/no " box," " menu" " box," " input" " box,"
.BR message " box," " text" " box," " info" " box, and"
.BR checklist " box."
.SH OPTIONS
.TP
.B \-\-clear
The screen will be cleared to the
.BR "screen attribute" " on exit."
.TP
.BI \-\-create-rc " file"
.RB "Since " dialog " supports run-time configuration,"
this can be used to dump a sample configuration file to the file specified
by
.IR file "."
.TP
.BI \-\-title " title"
Specifies a
.I title
string to be displayed at the top of the dialog box.
.TP
.B Box Options
.TP
.BI \-\-yesno " text height width"
.RB A " yes/no" " dialog box of size"
.I height
rows by
.I width
columns will be displayed. The string specified by
.I text
is displayed inside the dialog box. If this string is too long to be fitted
in one line, it will be automatically divided into multiple lines at
appropriate places. The
.I text
string may also contain the sub-string
.I
"\en"
or newline characters
.I `\en\'
to control line breaking explicitly.  This dialog box is useful for
asking questions that require the user to answer either yes or no.
.RB "The dialog box has a" " Yes" " button and a " No
button, in which the user can switch between by pressing the
.IR TAB " key."
.TP
.BI \-\-msgbox " text height width"
.RB A " message" " box is very similar to a" " yes/no" " box."
The only difference between a
.B message
box and a
.B yes/no
box is that a
.B message
box has only a single
.B OK
button. You can use this dialog box to display any message you like.
After reading the message, the user can press the
.I ENTER
key so that
.B dialog
will exit and the calling shell script can continue its operation.
.TP
.BI \-\-infobox " text height width"
.RB An " info" " box is basically a" " message" " box."
However, in this case,
.B dialog
will exit immediately after displaying the message to the user. The
screen is not cleared when
.B dialog
exits, so that the message will remain on the screen until the calling
shell script clears it later. This is useful when you want to inform
the user that some operations are carrying on that may require some
time to finish.
.TP
.BI \-\-inputbox " text height width"
.RB "An " input " box is useful when you want to ask questions that"
require the user to input a string as the answer. When inputing the
string, the
.I BACKSPACE
key can be used to correct typing errors. If the input string is longer than
can be fitted in the dialog box, the input field will be scrolled. On exit,
the input string will be printed on
.IR stderr "."
.TP
.BI \-\-textbox " file height width"
.RB A " text" " box lets you display the contents of a text file in a"
dialog box. It is like a simple text file viewer. The user can move
through the file by using the
.IR UP/DOWN ", " PGUP/PGDN
.RI and " HOME/END" " keys available on most keyboards."
If the lines are too long to be displayed in the box, the
.I LEFT/RIGHT
keys can be used to scroll the text region horizontally. For more
convenience, forward and backward searching functions are also provided.
.IP "\fB\-\-menu \fItext height width menu-height \fR[ \fItag item \fR] \fI..."
As its name suggests, a
.B menu
box is a dialog box that can be used to present a list of choices in
the form of a menu for the user to choose. Each menu entry consists of a
.IR tag " string and an " item " string. The"
.I tag
gives the entry a name to distinguish it from the other entries in the
menu. The
.I item
is a short description of the option that the entry represents. The
user can move between the menu entries by pressing the
.I UP/DOWN
keys, the first letter of the
.I tag
as a hot-key, or the number keys
.IR 1-9 ". There are"
.I menu-height
entries displayed in the menu at one time, but the menu will be
scrolled if there are more entries than that. When
.B dialog
exits, the
.I tag
of the chosen menu entry will be printed on
.IR stderr "."
.IP "\fB\-\-checklist \fItext height width list-height \fR[ \fItag item status \fR] \fI..."
.RB "A " checklist " box is similar to a " menu " box in that there are"
multiple entries presented in the form of a menu. Instead of choosing
one entry among the entries, each entry can be turned on or off by the
user. The initial on/off state of each entry is specified by
.IR status "."
On exit, a list of the
.I tag
strings of those entries that are turned on will be printed on
.IR stderr "."
.SH "RUN-TIME CONFIGURATION"
.TP 4
1.
Create a sample configuration file by typing:
.LP
.in +1i
"dialog --create-rc <file>"
.TP 4
2.
At start,
.B dialog
determines the settings to use as follows:
.RS
.TP 4
a)
if environment variable
.B DIALOGRC
is set, it's value determines the name of the configuration file.
.TP 4
b)
if the file in (a) can't be found, use the file
.I $HOME/.dialogrc
as the configuration file.
.TP 4
c)
if the file in (b) can't be found, use compiled in defaults.
.RE
.TP 4
3.
Edit the sample configuration file and copy it to some place that
.B dialog
can find, as stated in step 2 above.
.SH ENVIRONMENT
.TP 15
.B DIALOGRC
Define this variable if you want to specify the name of the configuration file
to use.
.SH FILES
.TP 20
.I $HOME/.dialogrc
default configuration file
.SH DIAGNOSTICS
Exit status is 0 if
.BR dialog " is exited by pressing the " Yes " or " OK
button, and 1 if the
.BR No " or " Cancel
button is pressed. Otherwise, if errors occur inside
.B dialog
or
.B dialog
is exited by pressing the
.I ESC
key, the exit status is -1.
.SH BUGS
Text files containing
.I tab
characters may cause problems with
.B text
box.
.I Tab
characters in text files must first be expanded to spaces before being
.RB "displayed by " text " box."
.TP
Screen update is too slow.
.SH AUTHOR
Savio Lam <lam836@cs.cuhk.hk>


ChangeLog
2007-10-06  Karl Berry  <karl@tug.org>

	* Build (TL_WORKDIR): allow override of "Work".

2007-10-02  Karl Berry  <karl@tug.org>

	* configure.in: remove sam2p special case, since we omitted the
	program.

2007-08-14  Joachim Schrod  <jschrod@acm.org>

	* configure.in: dvipdfm and dvipdfmx need pnglib and zlib, too.

2007-07-16  Karl Berry  <karl@tug.org>

	* Makefile.in (lib): new target, depends on @LIBSDEP@.
	(all): do not depend on @LIBSDEP@, redundant since we already
		descend and build.
	From Joachim, 28 Jun 2007 11:47:10.

2007-03-21  Karl Berry  <karl@tug.org>

	* Makefile.in (distclean): remove mention of klibtool.

2006-12-14  Karl Berry  <karl@tug.org>

	* Build: allow invoking from anywhere, and more overrides.

2006-12-11  Karl Berry  <karl@tug.org>

	* Makefile.in (world, install): hmm, maybe better as a dependency
	than a submake.

	* Makefile.in (install): reinstate strip.
	(strip): actually strip.

2006-12-08  Karl Berry  <karl@tug.org>

	* Build: set GNUMAKE to TL_MAKE for freetype2/configure, and
	export it.  Reported by Manfred and Vladimir.

2006-12-05  Karl Berry  <karl@tug.org>

	* Build: no tlbuild.log, since then the exit status is always
	successful.

2006-11-08  Karl Berry  <karl@tug.org>

	* configure.in (PKGS): remove sam2p after all.

2006-07-12  Karl Berry  <karl@tug.org>

	* utils/sam2p/Makefile (install): use cp -p, not -a.

2006-06-02  Karl Berry  <karl@tug.org>

	* texk/web2c/omegadir/com16bit.ch, alephdir/com16bit-rc2.ch
	(sup_main_memory, sup_save_size): increase per tex.ch.
	From Akira.


utils/dialog/dialog.c
/*
 *  dialog - Display simple dialog boxes from shell scripts
 *
 *  AUTHOR: Savio Lam (lam836@cs.cuhk.hk)
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 *
 *  HISTORY:
 *
 *  17/12/93 - Version 0.1 released.
 *
 *  19/12/93 - menu will now scroll if there are more items than can fit
 *             on the screen.
 *           - added 'checklist', a dialog box with a list of options that
 *             can be turned on or off. A list of options that are on is 
 *             returned on exit.
 *
 *  20/12/93 - Version 0.15 released.
 *
 *  29/12/93 - Incorporated patch from Patrick J. Volkerding
 *             (volkerdi@mhd1.moorhead.msus.edu) that made these changes:
 *             - increased MAX_LEN to 2048
 *             - added 'infobox', equivalent to a message box without pausing
 *             - added option '--clear' that will clear the screen
 *             - Explicit line breaking when printing prompt text can be
 *               invoked by real newline '\n' besides the string "\n"
 *           - an optional parameter '--title <string>' can be used to
 *             specify a title string for the dialog box
 *
 *  03/01/94 - added 'textbox', a dialog box for displaying text from a file.
 *           - Version 0.2 released.
 *
 *  04/01/94 - some fixes and improvements for 'textbox':
 *             - fixed a bug that will cause a segmentation violation when a
 *               line is longer than MAX_LEN characters. Lines will now be
 *               truncated if they are longer than MAX_LEN characters.
 *             - removed wrefresh() from print_line(). This will increase
 *               efficiency of print_page() which calls print_line().
 *             - display current position in the form of percentage into file.
 *           - Version 0.21 released.
 *
 *  05/01/94 - some changes for faster screen update.
 *
 *  07/01/94 - much more flexible color settings. Can use all 16 colors
 *             (8 normal, 8 highlight) of the Linux console.
 *
 *  08/01/94 - added run-time configuration using configuration file.
 *
 *  09/01/94 - some minor bug fixes and cleanups for menubox, checklist and
 *             textbox.
 *
 *  11/01/94 - added a man page.
 *
 *  13/01/94 - some changes for easier porting to other Unix systems (tested
 *             on Ultrix, SunOS and HPUX)
 *           - Version 0.3 released.
 * 
 *  08/06/94 - Patches by Stuart Herbert - S.Herbert@shef.ac.uk
 * 	       Fixed attr_clear and the textbox stuff to work with ncurses 1.8.5
 * 	       Fixed the wordwrap routine - it'll actually wrap properly now
 *	       Added a more 3D look to everything - having your own rc file could
 *	         prove 'interesting' to say the least :-)
 *             Added radiolist option
 *             Added backtitle option
 *	     - Version 0.4 released.
 *
 *  10/10/94 - Patches by Marc Ewing - marc@redhat.com
 *             Added --separate-output
 *             Added init val to inputbox
 *             Added guage widget
 */


#define __DIALOG_MAIN__


#include "dialog.h"
#ifdef HAVE_NCURSES
#include "colors.h"
#endif

static char *backtitle = NULL;
int separate_output = 0;

int main(int argc, char *argv[])
{
  int offset = 0, clear_screen = 0, end_common_opts = 0, retval;
  char *title = NULL;
  
#if defined(LOCALE)
  (void) setlocale(LC_ALL, "");
#endif

  if (argc < 2) {
    Usage(argv[0]);
    exit(-1);
  }
  else if (!strcmp(argv[1], "--create-rc")) {
#ifdef HAVE_NCURSES
    if (argc != 3) {
      Usage(argv[0]);
      exit(-1);
    }
    create_rc(argv[2]);
    return 0;
#else
    fprintf(stderr, "\nThis option is currently unsupported on your system.\n");
    return -1;
#endif
  }

  while (offset < argc-1 && !end_common_opts) {    /* Common options */
    if (!strcmp(argv[offset+1], "--title")) {
      if (argc-offset < 3 || title != NULL) {    /* No two "--title" please! */
        Usage(argv[0]);
        exit(-1);
      }
      else {
        title = argv[offset+2];
        offset += 2;
      }
    }
    else if (!strcmp(argv[offset+1], "--backtitle" ))
    {
      if (backtitle != NULL)
      {
        Usage(argv[0]);
        exit(-1);
      }
      else
      {
	backtitle = argv[offset+2];
	offset += 2;
      }
    }
    else if (!strcmp(argv[offset+1], "--separate-output" ))
    {
	separate_output = 1;
	offset++;
    }
    else if (!strcmp(argv[offset+1], "--clear")) {
      if (clear_screen) {    /* Hey, "--clear" can't appear twice! */
        Usage(argv[0]);
        exit(-1);
      }
      else if (argc == 2) {    /* we only want to clear the screen */
        init_dialog();
        refresh();    /* init_dialog() will clear the screen for us */
        endwin();
        return 0;
      }
      else {
        clear_screen = 1;
        offset++;
      }
    }
    else    /* no more common options */
      end_common_opts = 1;
  }

  if (argc-1 == offset) {    /* no more options */
    Usage(argv[0]);
    exit(-1);
  }

  /* Box options */

  if (!strcmp(argv[offset+1], "--yesno")) {
    if (argc-offset != 5) {
      Usage(argv[0]);
      exit(-1);
    }
    init_dialog();
    retval = dialog_yesno(title, argv[offset+2], atoi(argv[offset+3]),
                          atoi(argv[offset+4]));

    if (clear_screen) {    /* clear screen before exit */
      attr_clear(stdscr, LINES, COLS, screen_attr);
      refresh();
    }
    endwin();
    return retval;
  }
  else if (!strcmp(argv[offset+1], "--msgbox")) {
    if (argc-offset != 5) {
      Usage(argv[0]);
      exit(-1);
    }
    init_dialog();
    retval = dialog_msgbox(title, argv[offset+2], atoi(argv[offset+3]),
                           atoi(argv[offset+4]), 1);

    if (clear_screen) {    /* clear screen before exit */
      attr_clear(stdscr, LINES, COLS, screen_attr);
      refresh();
    }
    endwin();
    return retval;
  }
  else if (!strcmp(argv[offset+1], "--infobox")) {
    if (argc-offset != 5) {
      Usage(argv[0]);
      exit(-1);
    }
    init_dialog();
    retval = dialog_msgbox(title, argv[offset+2], atoi(argv[offset+3]),
                           atoi(argv[offset+4]), 0);

    if (clear_screen) {    /* clear screen before exit */
      attr_clear(stdscr, LINES, COLS, screen_attr);
      refresh();
    }
    endwin();
    return retval;
  }
  else if (!strcmp(argv[offset+1], "--textbox")) {
    if (argc-offset != 5) {
      Usage(argv[0]);
      exit(-1);
    }
    init_dialog();
    retval = dialog_textbox(title, argv[offset+2], atoi(argv[offset+3]),
                            atoi(argv[offset+4]));

    if (clear_screen) {    /* clear screen before exit */
      attr_clear(stdscr, LINES, COLS, screen_attr);
      refresh();
    }
    endwin();
    return retval;
  }
  else if (!strcmp(argv[offset+1], "--menu")) {
    if (argc-offset < 8 || ((argc-offset) % 2)) {
      Usage(argv[0]);
      exit(-1);
    }
    init_dialog();
    retval = dialog_menu(title, argv[offset+2], atoi(argv[offset+3]),
                         atoi(argv[offset+4]), atoi(argv[offset+5]),
                         (argc-offset-6)/2, argv+offset + 6);

    if (clear_screen) {    /* clear screen before exit */
      attr_clear(stdscr, LINES, COLS, screen_attr);
      refresh();
    }
    endwin();
    return retval;
  }
  else if (!strcmp(argv[offset+1], "--checklist")) {
    if (argc-offset < 9 || ((argc-offset-6) % 3)) {
      Usage(argv[0]);
      exit(-1);
    }
    init_dialog();
    retval = dialog_checklist(title, argv[offset+2], atoi(argv[offset+3]),
                              atoi(argv[offset+4]), atoi(argv[offset+5]),
                              (argc-offset-6)/3, argv+offset + 6);

    if (clear_screen) {    /* clear screen before exit */
      attr_clear(stdscr, LINES, COLS, screen_attr);
      refresh();
    }
    endwin();
    return retval;
  }
  else if (!strcmp(argv[offset+1], "--radiolist")) {
    if (argc-offset < 9 || ((argc-offset-6) % 3)) {
      Usage(argv[0]);
      exit(-1);
    }
    init_dialog();
    retval = dialog_radiolist(title, argv[offset+2], atoi(argv[offset+3]),
                              atoi(argv[offset+4]), atoi(argv[offset+5]),
                              (argc-offset-6)/3, argv+offset + 6);

    if (clear_screen) {    /* clear screen before exit */
      attr_clear(stdscr, LINES, COLS, screen_attr);
      refresh();
    }
    endwin();
    return retval;
  }
  else if (!strcmp(argv[offset+1], "--guage")) {
      if (argc-offset != 6) {
	  Usage(argv[0]);
	  exit(-1);
      }
      init_dialog();
      retval = dialog_guage(title,  argv[offset+2],
			    atoi(argv[offset+3]), atoi(argv[offset+4]),
			    atoi(argv[offset+5]));
      
      if (clear_screen) {    /* clear screen before exit */
	  attr_clear(stdscr, LINES, COLS, screen_attr);
	  refresh();
      }
      endwin();
      return retval;
  }
  else if (!strcmp(argv[offset+1], "--inputbox")) {
    if (argc-offset < 5 || argc-offset > 6) {
      Usage(argv[0]);
      exit(-1);
    }
    init_dialog();
    retval = dialog_inputbox(title, argv[offset+2], atoi(argv[offset+3]),
                             atoi(argv[offset+4]),
			     (argc-offset == 6) ? argv[offset+5] : (char *)NULL);

    if (clear_screen) {    /* clear screen before exit */
      attr_clear(stdscr, LINES, COLS, screen_attr);
      refresh();
    }
    endwin();
    return retval;
  }

  Usage(argv[0]);
  exit(-1);
}
/* End of main() */


/*
 * Print program usage
 */
void Usage(char *name)
{
  fprintf(stderr, "\
\ndialog version 0.3, by Savio Lam (lam836@cs.cuhk.hk).\
\n  patched to version %s by Stuart Herbert (S.Herbert@shef.ac.uk)\
\n\
\n* Display dialog boxes from shell scripts *\
\n\
\nUsage: %s --clear\
\n       %s --create-rc <file>\
\n       %s [--title <title>] [--separate-output] [--backtitle <backtitle>] [--clear] <Box options>\
\n\
\nBox options:\
\n\
\n  --yesno     <text> <height> <width>\
\n  --msgbox    <text> <height> <width>\
\n  --infobox   <text> <height> <width>\
\n  --inputbox  <text> <height> <width> [<init>]\
\n  --textbox   <file> <height> <width>\
\n  --menu      <text> <height> <width> <menu height> <tag1> <item1>...\
\n  --checklist <text> <height> <width> <list height> <tag1> <item1> <status1>...\
\n  --radiolist <text> <height> <width> <list height> <tag1> <item1> <status1>...\
\n  --guage     <text> <height> <width> <percent>\n", VERSION, name, name, name);
}
/* End of Usage() */


/*
 * Do some initialization for dialog
 */
void init_dialog(void)
{
  int i;
  
#ifdef HAVE_NCURSES
  if (parse_rc() == -1)    /* Read the configuration file */
    exit(-1);
#endif

  initscr();     /* Init curses */
  keypad(stdscr, TRUE);
  cbreak();
  noecho();

#ifdef HAVE_NCURSES
  if (use_colors || use_shadow)    /* Set up colors */
    color_setup();
#endif

  /* Set screen to screen attribute */
  attr_clear(stdscr, LINES, COLS, screen_attr);

  /* Display background title if it exists ... - SLH */
  if (backtitle != NULL)
  {
    wattrset(stdscr, screen_attr);
    wmove(stdscr, 0, 1);
    waddstr(stdscr,backtitle);
    wmove(stdscr, 1, 1);
    for(i=1; i<COLS-1; i++)
      waddch(stdscr, ACS_HLINE);
  }
  
  wnoutrefresh(stdscr);
}
/* End of init_dialog() */


#ifdef HAVE_NCURSES
/*
 * Setup for color display
 */
void color_setup(void)
{
  int i;

  if (has_colors()) {    /* Terminal supports color? */
    start_color();

    /* Initialize color pairs */
    for (i = 0; i < ATTRIBUTE_COUNT; i++)
      init_pair(i+1, color_table[i][0], color_table[i][1]);

    /* Setup color attributes */
    for (i = 0; i < ATTRIBUTE_COUNT; i++)
      attributes[i] = C_ATTR(color_table[i][2], i+1);
  }
}
/* End of color_setup() */
#endif


/*
 * Set window to attribute 'attr'
 */
void attr_clear(WINDOW *win, int height, int width, chtype attr)
{
/* ifdef HAVE_NCURSES
  wattrset(win, attr);    Set window to attribute 'attr'
  werase(win);
else */
  int i, j;

  wattrset(win, attr);    /* Set window to attribute 'attr' */
  for (i = 0; i < height; i++) {
    wmove(win, i, 0);
    for (j = 0; j < width; j++)
      waddch(win, ' ');
  }
  touchwin(win);
/* endif */
}
/* End of attr_clear() */


/*
 * Print a string of text in a window, automatically wrap around to the
 * next line if the string is too long to fit on one line. Note that the
 * string may contain "\n" to represent a newline character or the real
 * newline '\n', but in that case, auto wrap around will be disabled.
 */
void print_autowrap(WINDOW *win, char *prompt, int width, int y, int x)
{
  int first = 1, cur_x, cur_y;
  char tempstr[MAX_LEN+1], *word, *tempptr, *tempptr1;

  strcpy(tempstr, prompt);
  if ((strstr(tempstr, "\\n") != NULL) ||
      (strchr(tempstr, '\n') != NULL)) {    /* Prompt contains "\n" or '\n' */
    word = tempstr;
    cur_y = y;
    wmove(win, cur_y, x);
    while (1) {
      tempptr = strstr(word, "\\n");
      tempptr1 = strchr(word, '\n');
      if (tempptr == NULL && tempptr1 == NULL)
        break;
      else if (tempptr == NULL) {    /* No more "\n" */
        tempptr = tempptr1;
        tempptr[0] = '\0';
      }
      else if (tempptr1 == NULL) {    /* No more '\n' */
        tempptr[0] = '\0';
        tempptr++;
      }
      else {    /* Prompt contains both "\n" and '\n' */
        if (strlen(tempptr)-2 < strlen(tempptr1)-1) {
          tempptr = tempptr1;
          tempptr[0] = '\0';
        }
        else {
          tempptr[0] = '\0';
          tempptr++;
        }
      }

      waddstr(win, word);
      word = tempptr + 1;
      wmove(win, ++cur_y, x);
    }
    waddstr(win, word);
  }
  else if (strlen(tempstr) <= width-x*2) {    /* If prompt is short */
    wmove(win, y, (width - strlen(tempstr)) / 2);
    waddstr(win, tempstr);
  }
  else {
    cur_x = x;
    cur_y = y;
    /* Print prompt word by word, wrap around if necessary */
    while ((word = strtok(first ? tempstr : NULL, " ")) != NULL) {
      if (first)    /* First iteration */
        first = 0;
      if (cur_x+strlen(word) > width) {    /* wrap around to next line */
        cur_y++;
        cur_x = x;
      }
      wmove(win, cur_y, cur_x);
      waddstr(win, word);
      getyx(win, cur_y, cur_x);
      cur_x++;
    }
  }
}
/* End of print_autowrap() */


/*
 * Print a button
 */
void print_button(WINDOW *win, char *label, int y, int x, int selected)
{
  int i, temp;

  wmove(win, y, x);
  wattrset(win, selected ? button_active_attr : button_inactive_attr);
  waddstr(win, "<");
  temp = strspn(label, " ");
  label += temp;
  wattrset(win, selected ? button_label_active_attr : button_label_inactive_attr);
  for (i = 0; i < temp; i++)
    waddch(win, ' ');
  wattrset(win, selected ? button_key_active_attr : button_key_inactive_attr);
  waddch(win, label[0]);
  wattrset(win, selected ? button_label_active_attr : button_label_inactive_attr);
  waddstr(win, label+1);
  wattrset(win, selected ? button_active_attr : button_inactive_attr);
  waddstr(win, ">");
  wmove(win, y, x+temp+1);
}
/* End of print_button() */


/*
 * Draw a rectangular box with line drawing characters
 */
void draw_box(WINDOW *win, int y, int x, int height, int width, chtype box, chtype border)
{
  int i, j;

  wattrset(win, 0);
  for (i = 0; i < height; i++) {
    wmove(win, y + i, x);
    for (j = 0; j < width; j++)
      if (!i && !j)
        waddch(win, border | ACS_ULCORNER);
      else if (i == height-1 && !j)
        waddch(win, border | ACS_LLCORNER);
      else if (!i && j == width-1)
        waddch(win, box | ACS_URCORNER);
      else if (i == height-1 && j == width-1)
        waddch(win, box | ACS_LRCORNER);
      else if (!i)
        waddch(win, border | ACS_HLINE);
      else if (i == height-1)
        waddch(win, box | ACS_HLINE);
      else if (!j)
        waddch(win, border | ACS_VLINE);
      else if (j == width-1)
        waddch(win, box | ACS_VLINE);
      else
        waddch(win, box | ' ');
  }
}
/* End of draw_box() */


#ifdef HAVE_NCURSES
/*
 * Draw shadows along the right and bottom edge to give a more 3D look
 * to the boxes
 */
void draw_shadow(WINDOW *win, int y, int x, int height, int width)
{
  int i;

  if (has_colors()) {    /* Whether terminal supports color? */
    wattrset(win, shadow_attr);
    wmove(win, y + height, x + 2);
    for (i = 0; i < width; i++)
      waddch(win, winch(win) & A_CHARTEXT);
    for (i = y + 1; i < y + height + 1; i++) {
      wmove(win, i, x + width);
      waddch(win, winch(win) & A_CHARTEXT);
      waddch(win, winch(win) & A_CHARTEXT);
    }
    wnoutrefresh(win);
  }
}
/* End of draw_shadow() */
#endif
